### PR TITLE
Include Channel’s ExternalName services in Istio mesh

### PR DESCRIPTION
**Pull request by pierDipi**
Implementation for https://docs.google.com/document/d/1Y9BukjzUxfl920KltVtH4L5gta0pHOt5NEsLqdgZNY4/edit

For review I'd recommend checking out the PR locally

To enabled the integration you need to specify in `config-features` an `istio` flag:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-features
  namespace: knative-eventing
data:
  istio: "enabled"
```

### CODE REVIEW

**Comment by pierDipi** [comment]
Waiting for https://github.com/knative-sandbox/eventing-kafka-broker/pull/3045

**Comment by pierDipi** [comment]
Waiting for https://github.com/knative/eventing/pull/6862

**Comment by matzew** [comment]
```
Error: error processing import paths in "/home/matzew/go/src/knative.dev/eventing-istio/config/eventing-istio/controller/500-controller.yaml": error resolving image references: Post "https://index.docker.io/v2/matzew/controller-7bf9c28a196da918a0e226f1cbddf82e/blobs/uploads/?from=static&mount=sha256%3A6428370c8e2ad3036b85fb37805c793508e0f06b3a7220f5dae45b08d5c531a7&origin=distroless.dev": GET https://auth.docker.io/token?scope=repository%3Amatzew%2Fcontroller-7bf9c28a196da918a0e226f1cbddf82e%3Apull%2Cpush+repository%3Astatic%3Apull&scope=repository%3Amatzew%2Fcontroller-7bf9c28a196da918a0e226f1cbddf82e%3Apush%2Cpull&service=registry.docker.io: unexpected status code 400 Bad Request: {"details":"invalid repository name"}
```

from `hack/instsall.sh`  Looks like some "build/ko" is needed?

**Comment by pierDipi** [comment]
> ```
> Error: error processing import paths in "/home/matzew/go/src/knative.dev/eventing-istio/config/eventing-istio/controller/500-controller.yaml": error resolving image references: Post "https://index.docker.io/v2/matzew/controller-7bf9c28a196da918a0e226f1cbddf82e/blobs/uploads/?from=static&mount=sha256%3A6428370c8e2ad3036b85fb37805c793508e0f06b3a7220f5dae45b08d5c531a7&origin=distroless.dev": GET https://auth.docker.io/token?scope=repository%3Amatzew%2Fcontroller-7bf9c28a196da918a0e226f1cbddf82e%3Apull%2Cpush+repository%3Astatic%3Apull&scope=repository%3Amatzew%2Fcontroller-7bf9c28a196da918a0e226f1cbddf82e%3Apush%2Cpull&service=registry.docker.io: unexpected status code 400 Bad Request: {"details":"invalid repository name"}
> ```
> 
> from `hack/instsall.sh` Looks like some "build/ko" is needed?

Yes

**Comment by matzew** [comment]
should the script be updated? 

I just ran:
hack/install-deps.sh
hack/install.sh

File: cmd/controller/main.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		log.Printf("Readiness and health check server listening on port %s", port)
+
+		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Fatal(err)
+		}
+	}()
+
+	ctx = istiofilteredfactory.WithSelectors(ctx, service.IstioResourceSelector)
+
+	sharedmain.MainWithContext(ctx, "eventing-istio-controller",
```
Comment (matzew) [inline cmd/controller/main.go:66]: Longer term: we can integrate this w/ the "pure"/"core" `eventing-controller`?

Comment (pierDipi) [reply cmd/controller/main.go:66]: I don't think we are planning to do that

File: pkg/reconciler/service/controller.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+			return false
+		}
+
+		logger.Debug("Filtering Service",
+			zap.String("namespace", svc.GetNamespace()),
+			zap.String("name", svc.GetName()),
+			zap.Any("labels", svc.GetLabels()),
+		)
+
+		if svc.Spec.ExternalName == "" {
```
Comment (matzew) [inline pkg/reconciler/service/controller.go:132]: when is this the case? Any

Comment (matzew) [reply pkg/reconciler/service/controller.go:132]: any comment (sorry, the browser froze...)

**Comment by matzew** [comment]
Ran some `KafkaChannel` / `Broker` tests:


My broker:
```
k get brokers.eventing.knative.dev -n test 
NAME      URL                                                                     AGE   READY   REASON
default   http://broker-ingress.knative-eventing.svc.cluster.local/test/default   31m   True    
```

That points to a kafka channel CFG, like:
```yaml
spec:
  config:
    apiVersion: v1
    kind: ConfigMap
    name: kafka-channel
    namespace: knative-eventing
```

Here is the backing channel:

```
k get kafkachannels.messaging.knative.dev -n test 
NAME                  READY   REASON   URL                                                            AGE
default-kne-trigger   True             http://default-kne-trigger-kn-channel.test.svc.cluster.local   32m
```

Which has a service

```
k get services -n test 
NAME                             TYPE           CLUSTER-IP   EXTERNAL-IP                                                PORT(S)   AGE
default-kne-trigger-kn-channel   ExternalName   <none>       kafka-channel-ingress.knative-eventing.svc.cluster.local   80/TCP    32m
```

and now I see the `destinationrule`:

```
k get destinationrules.networking.istio.io -n test                                      
NAME                             HOST                                                    AGE
default-kne-trigger-kn-channel   default-kne-trigger-kn-channel.test.svc.cluster.local   32m
```

and here is the details, and I see that the "service reconciler" did create it:

```
k get destinationrules.networking.istio.io -n test default-kne-trigger-kn-channel -oyaml 
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  creationTimestamp: "2023-05-04T08:01:06Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: knative-eventing-istio-controller
    messaging.knative.dev/role: kafka-channel
  name: default-kne-trigger-kn-channel
  namespace: test
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Service
    name: default-kne-trigger-kn-channel
    uid: f31eb2a1-79d0-4716-a9ad-da9e4aa6d495
  resourceVersion: "123970"
  uid: f7752f47-216a-44c5-ab36-68df81a907d0
spec:
  host: default-kne-trigger-kn-channel.test.svc.cluster.local
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
```

**Comment by matzew** [comment]
https://github.com/pierDipi/eventing-istio/blob/channel-reconciler/pkg/reconciler/service/service.go#L60-L64

The comment here is slightly different than the log. Wanna adjust?

**Comment by matzew** [comment]
/lgtm
/approve

/hold
(feel free to unhold)

**Comment by pierDipi** [comment]
> https://github.com/pierDipi/eventing-istio/blob/channel-reconciler/pkg/reconciler/service/service.go#L60-L64
> 
> The comment here is slightly different than the log. Wanna adjust?

Done

File: pkg/reconciler/service/controller.go (hunk 2 above)
Comment (pierDipi) [reply pkg/reconciler/service/controller.go:132]: This is double checking that we're not queue-ing services which aren't what we expect, mostly robustness

**Review by matzew — APPROVED** [review]
/lgtm
/approve

**Review by creydr — COMMENTED** [review]
LGTM from my side. Leaving it open for others to review too

**Comment by pierDipi** [comment]
@creydr do you have any comments?

**Comment by pierDipi** [comment]
lol :)

**Comment by pierDipi** [comment]
/override style / Golang / Boilerplate Check (go)

**Pull request merged**